cas/ctg_store.c

```diff
@@ -827,40 +974,35 @@ static bool ctg_is_ordinary(const struct m0_cas_ctg *ctg)
 
 static bool ctg_op_cb(struct m0_clink *clink)
 {
-	struct m0_ctg_op *ctg_op   = M0_AMB(ctg_op, clink, co_clink);
-	struct m0_be_op  *op      = ctg_beop(ctg_op);
-	int               opc     = ctg_op->co_opcode;
-	int               ct      = ctg_op->co_ct;
-	struct m0_be_tx  *tx      = &ctg_op->co_fom->fo_tx.tx_betx;
-	void             *arena   = ctg_op->co_anchor.ba_value.b_addr;
-	struct m0_buf     cur_key = {};
-	struct m0_buf     cur_val = {};
-	struct m0_chan   *ctg_chan = &ctg_op->co_ctg->cc_chan.bch_chan;
-	struct m0_buf    *dst;
-	struct m0_buf    *src;
-	int               rc;
+	struct m0_ctg_op  *ctg_op   = M0_AMB(ctg_op, clink, co_clink);
+	struct m0_be_op   *op       = ctg_beop(ctg_op);
+	int                opc      = ctg_op->co_opcode;
+	int                ct       = ctg_op->co_ct;
+	struct m0_be_tx   *tx       = &ctg_op->co_fom->fo_tx.tx_betx;
+	struct m0_chan    *ctg_chan = &ctg_op->co_ctg->cc_chan.bch_chan;
+	int                rc;
+	struct m0_cas_ctg *ctg;
+	struct fid_key    *fk;
+	struct meta_value *mv;
 
 	if (op->bo_sm.sm_state != M0_BOS_DONE)
 		return true;
 
+	/* Versioned API is synchronous. */
```

Comment:
@ivan-alekhin 

In my understanding, versioned operations are used in DTM0 recovery/DIX repair.
No need to make it async.

Reply:
Let me do a little bit of nitpicking on my own comment.

First of all, the statement `Versioned API is synchronous. ` is wrong. `The versioned API implementation is synchronous.` is the right statement. The normal CAS and the versioned CAS are exposed through the same API. The API supports async behavior (co_fom gets awaken when ctg_beop gets triggered). The API is implemented on the top of a mostly synchronous btree. Both behaviors (versioned and normal) are, essentially, synchronous. The only difference here is that normal CAS has some code that may help it to work on the top of an asynchronous btree. Versioned CAS does not have such code.

The second thing is the question "why is the implementation synchronous?". Versioned CAS cannot be async because asynchronous versioned CAS requires asynchronous btree implementation which does not exist yet. It would also require modifications at the CAS FOMs level to support chained async calls (for example, lookup+save or a series of cursor_next calls). Moreover, the normal CAS is also synchronous because it uses the same synchronous btree. However, normal CAS has an opportunity to become asynchronous. Versioned CAS is simpler, and it does not have such an opportunity.

Reply:
OK. Let me update the comments.

Reply:
Fixed.